### PR TITLE
gpcheckcat should report duplicate index violations

### PR DESCRIPTION
**Pull request by kaknikhil**
As a follow up to #621 , this pr adds a new check to gpcheckcat which will report any duplicate references for all the unique catalog indexes

- The code uses a function from indexscan to read the index values . We had to modify the return value to optionally not include the actual column values.
- Add behave and unit tests for the python module
- Add ICG test for the readindex function

Example of violation on master 

```
 select * from readindex('pg_compression_compname_index'::regclass) as (ictid tid, hctid tid, aotid text,istatus text, hstatus text);
 ictid | hctid | aotid | istatus |           hstatus
 -------+-------+-------+---------+-----------------------------
 (1,1) | (0,4) | N/A   | USED    | XMIN_COMITTTED XMAX_INVALID
 (1,2) | (0,4) | N/A   | USED    | XMIN_COMITTTED XMAX_INVALID
```

Authors:
Nikhil Kak , Chris Hajas, Lawrence Hamel

### CODE REVIEW

**Comment by kaknikhil** [comment]
@asimrp : can you please review ?

File: gpMgmt/bin/gpcheckcat [hunk 1]
```diff
@@ -3403,41 +3403,66 @@ def duplicatePersistentEntryQuery(catname, pkey, excol, state):
     return qry
 
 # -------------------------------------------------------------------------------
-def checkUniqueIndexViolation():
+
+
+def log_begin_unique_index_violation():
```
Comment (larham) [inline gpMgmt/bin/gpcheckcat]: Consider adding __ to beginning of name to clarify that this is a private method

File: gpMgmt/bin/gpcheckcat [hunk 2]
```diff
@@ -3403,41 +3403,66 @@ def duplicatePersistentEntryQuery(catname, pkey, excol, state):
...
+    except Exception, e:
+        setError(ERROR_NOREPAIR)
+        myprint('  Execution error: ' + str(e))
+
+    check_name = 'index duplication on unique indexes'
+    log_end_unique_index_violation(check_name, violations)
 
-    logger.error('\n'.join(log_output))
+
+def log_end_unique_index_violation(check_name, violations):
```
Comment (larham) [inline gpMgmt/bin/gpcheckcat]: consider adding __ to beginning of method name to emphasize private-ness

File: gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py [hunk 3]
```diff
@@ -49,37 +50,65 @@ def test_running_unknown_check__raises_exception(self):
...
+        self.unique_index_violation_check.run_check_table_duplication.assert_called_with(self.db_connection)
 
-    def test_running_unique_index_violation_check__when_no_violations_are_found__passes_the_check(self):
-        self.subject.runOneCheck('unique_index_violation')
+        self.assertTrue(self.subject.GV.checkStatus)
+        self.subject.setError.assert_not_called()
+
+    def test_running_unique_index_violation_table_duplication_check__when_violations_are_found__fails_the_check(self):
+        self.unique_index_violation_check.run_check_table_duplication.return_value = [
+            dict(table_oid=123, table_name='stephen_table', index_name='finger', column_names='c1, c2', violated_segments=[-1,8]),
```
Comment (larham) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py]: This reminds me: Consider giving Stephen Wu an authorship credit

Comment (kaknikhil) [inline gpMgmt/bin/gppylib/test/unit/test_unit_gpcheckcat.py]: stephen added the test_running_unique_index_violation_check__when_no_violations_are_found__passes_the_check in pr #621 . i am just renaming it here

File: contrib/indexscan/indexscan.c [hunk 4]
```diff
@@ -154,6 +154,18 @@ readindex(PG_FUNCTION_ARGS)
 	if (SRF_IS_FIRSTCALL())
 	{
 		Oid		irelid = PG_GETARG_OID(0);
+
+		/* By default we want readindex to return everything. Optionally use the boolean value to skip the corresponding column values from the base table
```
Comment (ashwinstar) [inline contrib/indexscan/indexscan.c]: Comments need to follow multiline style 
```
/*
 *
 *
 */ 
```
Plus, also fit within 80 column width.

File: contrib/indexscan/indexscan.c [hunk 5]
```diff
@@ -154,6 +154,18 @@ readindex(PG_FUNCTION_ARGS)
...
+		 for ex both are valid:
+		 CREATE OR REPLACE FUNCTION readindex(oid) RETURNS SETOF record AS '$libdir/indexscan', 'readindex' LANGUAGE C STRICT;  -- returns everything
+		 CREATE OR REPLACE FUNCTION readindex(oid, bool) RETURNS SETOF record AS '$libdir/indexscan', 'readindex' LANGUAGE C STRICT; -- skips the corresponding values if the 2nd arg is true
+		*/
+		bool skipValues = false;
```
Comment (ashwinstar) [inline contrib/indexscan/indexscan.c]: Missing to guess so would be great to mention the motivation behind why wish to skip the values ?
The current code change doesn't seem right me, we seems skipping only creation of tupledesc but the actual fetch from heaptuple and will all happen.

**Comment by kaknikhil** [comment]
after discussion with @ashwinstar , closing this pull request because we want to refactor the readindex code before using it in gpcheckcat.

**Comment by asimrp** [comment]
Sorry for the delated response.  The intent here is to catch uniqueness violations across the cluster.  The query run by gpcheckcat would fetch results of `readindex()` from all segments as well as master and find repeated tuples.  But if we leave out the index key in `readindex()` output, as done by this PR, we may miss out repetitions of some kind.  E.g. consider a corrupted unique index having two entires pointing to two different heap TIDs.  But value of the index key is the same for the two entires.  This PR will fail to catch such duplicates, because all fields except the index key column(s) differ.

**Pull request closed without merging**